File: algorithms/Non-overlapping Intervals/Non-overlapping Intervals.cpp
class Solution {
public:
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        priority_queue<pair<int, int> > pq;
        int ans = 0;
        //maxheap by ending time
        pq.push({intervals[0][1], intervals[0][0]});
        for(int i=1;i<intervals.size();i++){
            //if start of this interval is overlapping with max end time of some prev interval
            if(pq.top().first > intervals[i][0]){
                //remove one with max ending time
                if(pq.top().first > intervals[i][1]){
                    pq.pop();
                    pq.push({intervals[i][1], intervals[i][0]});
                }
                //update ans
                ans++;
            }
            else{
                //push current interval.
                pq.push({intervals[i][1], intervals[i][0]});
            }
        }
        return ans;
    }
};

File: algorithms/Non-overlapping Intervals/Non-overlapping Intervals.java
// |-------|
//   |--|

// |-------|
//.   |-------|

// |-------|
//.           |-------|

class Solution {
    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a,b) -> a[0] - b[0]);
        
        int start = intervals[0][0];
        int end = intervals[0][1];
        int res = 0;
        
        for (int i = 1; i < intervals.length; i++){
            int[] interval = intervals[i];
            
            if(interval[0] >= start && interval[0] < end){
                res++;
                if (interval[1] >= end)
                    continue;
            }
            start = interval[0];
            end = interval[1];
    }
        
        return res;
    }
}
